bn_mp_fwrite.c

```diff
@@ -8,29 +8,19 @@ mp_err mp_fwrite(const mp_int *a, int radix, FILE *stream)
 {
    char *buf;
    mp_err err;
-   int len;
+   size_t length;
 
-   if ((err = mp_radix_size(a, radix, &len)) != MP_OKAY) {
-      return err;
-   }
-
-   buf = (char *) MP_MALLOC((size_t)len);
-   if (buf == NULL) {
-      return MP_MEM;
-   }
-
-   if ((err = mp_to_radix(a, buf, (size_t)len, radix)) != MP_OKAY) {
+   if ((err = mp_to_radix(a, &buf, &length, radix)) != MP_OKAY) {
       goto LBL_ERR;
    }
-
-   if (fwrite(buf, (size_t)len, 1uL, stream) != 1uL) {
+   if (fwrite(buf, length, 1uL, stream) != 1uL) {
       err = MP_ERR;
       goto LBL_ERR;
    }
-   err = MP_OKAY;
 
 LBL_ERR:
-   MP_FREE_BUFFER(buf, (size_t)len);
+   /* "length" can be smaller than the allocated buffer! */
```

Comment:
is that a problem @minad ?

as of https://github.com/libtom/libtommath/pull/192#issue-267356213 it should be a problem.

Reply:
Yes, that can be a showstopper.

That's one of the two reasons I waited the whole WE instead of the obligatory 24h but @minad seeems to enjoy his hiking tour a bit longer. (Yes, I admit, I am a bit envious ;-) )

The other one is that it is not possible anymore to use your own chunk of memory. I think @minad dropped a hint or two regarding static arrays, IIRC?

Reply:
Sorry I don't have much time right now. But from what I see this looks very much like a problem :)